src/core/net/dns_client.cpp

```diff
@@ -1077,21 +1196,219 @@ void Client::HandleTimer(void)
+
+    Message              *message   = nullptr;
+    size_t                totalRead = 0;
+    size_t                offset    = 0;
+    const otLinkedBuffer *data;
+
+    if (aEndOfStream)
+    {
+        // Cleanup is done in dissconnected callback.
+        IgnoreError(mEndpoint.SendEndOfStream());
+    }
+
+    SuccessOrExit(mEndpoint.ReceiveByReference(data));
+    VerifyOrExit(data != nullptr);
+
+    message = mSocket.NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    while (aBytesAvailable > totalRead)
+    {
+        uint16_t length;
+
+        // Read the 'length' field.
+        SuccessOrExit(ReadFromLinkBuffer(data, offset, *message, sizeof(uint16_t)));
+
+        IgnoreError(message->Read(/* aOffset */ 0, length));
+        length = HostSwap16(length);
+
+        // Try to read `length` bytes.
+        IgnoreError(message->SetLength(0));
+        SuccessOrExit(ReadFromLinkBuffer(data, offset, *message, length));
+
+        totalRead += length + sizeof(uint16_t);
+
+        // Now process the read message as query response.
+        {
+            Response  response;
+            QueryType type;
+            Error     responseError;
+
+            response.mInstance = &Get<Instance>();
+            response.mMessage  = message;
+
+            if (ParseResponse(response, type, responseError) == kErrorNone)
+            {
+                if (responseError == kErrorNone && length > OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_QUERY_MAX_SIZE)
+                {
+                    LogWarn("Dns query over wasn't received message too big.");
+                    responseError = kErrorNoBufs;
+                }
+                FinalizeQuery(response, type, responseError);
+            }
+        }
+
+        IgnoreError(message->SetLength(0));
+
+        // Loop again to see if we can read another response.
+    }
+
+exit:
+    // Inform `mEndPoint` about the total read and processed bytes
+    IgnoreError(mEndpoint.CommitReceive(totalRead, /* aFlags */ 0));
+    FreeMessage(message);
+}
+
+void Client::HandleTcpReceiveAvailableCallback(otTcpEndpoint *aEndpoint,
+                                               size_t         aBytesAvailable,
+                                               bool           aEndOfStream,
+                                               size_t         aBytesRemaining)
+{
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))
+        ->HandleTcpReceiveAvailable(aEndpoint, aBytesAvailable, aEndOfStream, aBytesRemaining);
+}
+
+void Client::HandleTcpDisconnected(otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    Query    *query;
+    QueryInfo info;
+
+    mEndpoint.Deinitialize();
+    mTcpState = kTcpUninitialized;
+    VerifyOrExit(aReason != OT_TCP_DISCONNECTED_REASON_NORMAL);
+
+    // Abort queries in case of connection failures
+    for (Message &msg : mQueries)
+    {
+        query = static_cast<Query *>(&msg);
```

Comment:
We can define `Query &query` as loop param directly (no need to cast from `Message` to `Query`).
```suggestion
    for (Query &query : mQueries)
    {
```

Reply:
fixed.